src/db.c

```diff
@@ -248,8 +248,9 @@ void dbOverwrite(redisDb *db, robj *key, robj *val) {
  * All the new keys in the database should be created via this interface.
  * The client 'c' argument may be set to NULL if the operation is performed
  * in a context where there is no clear client performing the operation. */
-void genericSetKey(client *c, redisDb *db, robj *key, robj *val, int keepttl, int signal) {
-    if (lookupKeyWrite(db,key) == NULL) {
+void genericSetKeyLookup(client *c, redisDb *db, robj *key, robj *val,int keepttl,
+                         int signal, int keyLooked, int keyFound) {
+    if ((!keyLooked || !keyFound) && lookupKeyWrite(db,key) == NULL) {
         dbAdd(db,key,val);
```

Comment:
I think it might be more straightforward to pass the lookup result through to this.
```suggestion
void genericSetKeyWithoutLookup(client *c, redisDb *db, robj *key, robj *val,int keepttl,
                         int signal, robj *key) {
    if (key == NULL) {
        dbAdd(db,key,val);
```

Reply:
1. I was suggesting having two functions, one that does the lookup and passes in the value and the second that accepts the passed in value that was looked up externally. 
2. I suppose it could just be a single bool flag as well, we don't necessarily need to pass around the Robj *.

Reply:
i now read all your comments again... if we always call  lookupKeyWrite(db,key) outside the genericSetKeyWithoutLookup, we don't really need any boolean. its a ggod idea, and much clearer ... I will post a pr soon